lib/services/service-scaffold.ts

```diff
@@ -0,0 +1,217 @@
+ * - An event registration and handling standard.
+ * - Checks for and extracts data from context.
+ * - A standard way of getting contextualised workers.
+ *
+ * In exchange you agree to:
+ * - provide a `emitData` method which takes a ServiceEmitContext and returns a promise.
+ * - provide a `verify` function which should check that incoming data isn't spoofed.
+ * - provide a `serviceName` getter, which because it is based on file path cannot be inherited away.
+ * - provide a `apiHandle` getter, which should return the underlying object that executes the requests.
+ * - enqueue your events with `context` and `event`, as per `ServiceScaffoldServiceEvent`.
+ */
+export abstract class ServiceScaffold<T> extends WorkerClient<T> implements ServiceListener, ServiceEmitter {
+	/** A singleton express instance for all web-hook based services to share. */
+	private static singletonExpressApp: express.Express;
+
+	private instanceExpressApp?: express.Express;
+
+	/** A place to put output for debug and reference. */
+	private loggerInstance = new Logger();
+
+	/** Store a list of actions to perform when particular actions happen */
+	private eventListeners: { [event: string]: ServiceRegistration[] } = {};
+
+	constructor(expressApp?: express.Express) {
+		super();
+		this.instanceExpressApp = expressApp;
+	}
+
+	/**
+	 * Store an event of interest, so that the method gets triggered appropriately.
+	 * @param registration  Registration object with event trigger and other details.
+	 */
+	public registerEvent(registration: ServiceRegistration): void {
+		// Store each event registration in an object of arrays.
+		for (const event of registration.events) {
+			if (!this.eventListeners[event]) {
+				this.eventListeners[event] = [];
+			}
+			this.eventListeners[event].push(registration);
+		}
+	}
+
+	/**
+	 * Emit data to the service.
+	 * @param data  Service Emit Request to send, if relevant.
+	 * @returns     Details of the successful transmission from the service.
+	 */
+	public sendData(data: ServiceEmitRequest): Promise<ServiceEmitResponse> {
+		const context = data.contexts[this.serviceName] as ServiceEmitContext;
+		if (!context) {
+			return Promise.resolve({
+				err: new ProcBotError(ProcBotErrorCode.EmitterContextAbsent, `No ${this.serviceName} context`),
+				source: this.serviceName,
+			});
+		}
+
+		return this.emitData(context)
+		.then((response: any) => {
+			return {
+				response,
+				source: this.serviceName,
+			};
+		})
+		.catch((err: TypedError) => {
+			return {
+				err,
+				source: this.serviceName,
+			};
+		});
+	}
+
+	/**
+	 * Queue an event ready for running in a child, here to type guard.
+	 * @param data  The WorkerEvent to add to the queue for processing.
+	 */
+	protected queueData = (data: ServiceScaffoldServiceEvent) => {
+		if (this.verify(data)) {
+			super.queueEvent({
+				data,
+				workerMethod: this.handleEvent
+			});
+		} else {
+			this.logger.log(LogLevel.WARN, `Event failed verification.`);
```

Comment:
Probably worth giving more details on this, if they can be extracted from the data.